In [15]:
import pandas as pd
import numpy as np


Import Numpy and Pandas libraries for reading csv files and to perform analysis

In [2]:
df = pd.read_csv('garments_worker_productivity.csv')
df.head()

date   quarter  department       day  team  targeted_productivity  \
0  1/1/2015  Quarter1      sweing  Thursday     8                   0.80   
1  1/1/2015  Quarter1  finishing   Thursday     1                   0.75   
2  1/1/2015  Quarter1      sweing  Thursday    11                   0.80   
3  1/1/2015  Quarter1      sweing  Thursday    12                   0.80   
4  1/1/2015  Quarter1      sweing  Thursday     6                   0.80   

     smv     wip  over_time  incentive  idle_time  idle_men  \
0  26.16  1108.0       7080         98        0.0         0   
1   3.94     NaN        960          0        0.0         0   
2  11.41   968.0       3660         50        0.0         0   
3  11.41   968.0       3660         50        0.0         0   
4  25.90  1170.0       1920         50        0.0         0   

   no_of_style_change  no_of_workers  actual_productivity  
0                   0           59.0             0.940725  
1                   0            8.0             0.886500  
2                   0           30.5             0.800570  
3                   0           30.5             0.800570  
4                   0           56.0             0.800382

Reads the CSV file garments_worker_productivity.csv into a pandas DataFrame called df, which stores the data in a table format

In [3]:
df.shape

(1197, 15)

The df.shape function returns the dimensions of the DataFrame, showing the number of rows and columns as a tuple.(1197, 15) means the DataFrame has 1197 rows and 15 columns.

In [4]:
df.isnull().sum()


date                       0
quarter                    0
department                 0
day                        0
team                       0
targeted_productivity      0
smv                        0
wip                      506
over_time                  0
incentive                  0
idle_time                  0
idle_men                   0
no_of_style_change         0
no_of_workers              0
actual_productivity        0
dtype: int64

df.isnull().sum() shows the number of missing (NaN) values in each column of the DataFrame

In [5]:
df.describe()


team  targeted_productivity          smv           wip  \
count  1197.000000            1197.000000  1197.000000    691.000000   
mean      6.426901               0.729632    15.062172   1190.465991   
std       3.463963               0.097891    10.943219   1837.455001   
min       1.000000               0.070000     2.900000      7.000000   
25%       3.000000               0.700000     3.940000    774.500000   
50%       6.000000               0.750000    15.260000   1039.000000   
75%       9.000000               0.800000    24.260000   1252.500000   
max      12.000000               0.800000    54.560000  23122.000000   

          over_time    incentive    idle_time     idle_men  \
count   1197.000000  1197.000000  1197.000000  1197.000000   
mean    4567.460317    38.210526     0.730159     0.369256   
std     3348.823563   160.182643    12.709757     3.268987   
min        0.000000     0.000000     0.000000     0.000000   
25%     1440.000000     0.000000     0.000000     0.000000   
50%     3960.000000     0.000000     0.000000     0.000000   
75%     6960.000000    50.000000     0.000000     0.000000   
max    25920.000000  3600.000000   300.000000    45.000000   

       no_of_style_change  no_of_workers  actual_productivity  
count         1197.000000    1197.000000          1197.000000  
mean             0.150376      34.609858             0.735091  
std              0.427848      22.197687             0.174488  
min              0.000000       2.000000             0.233705  
25%              0.000000       9.000000             0.650307  
50%              0.000000      34.000000             0.773333  
75%              0.000000      57.000000             0.850253  
max              2.000000      89.000000             1.120437

df.describe() provides a statistical summary of the DataFrame, including count, mean, standard deviation, minimum, maximum, and quartiles for each numeric column.

In [6]:
df.dtypes


date                      object
quarter                   object
department                object
day                       object
team                       int64
targeted_productivity    float64
smv                      float64
wip                      float64
over_time                  int64
incentive                  int64
idle_time                float64
idle_men                   int64
no_of_style_change         int64
no_of_workers            float64
actual_productivity      float64
dtype: object

df.dtypes displays the data types of each column in the DataFrame. This helps you understand the kind of data each column holds, such as int64, float64, or object

In [11]:
# Select only numeric columns
numeric_df = df.select_dtypes(include=[np.number])

# Calculate correlation matrix
correlation_matrix = numeric_df.corr()

# Display the correlation matrix
correlation_matrix

team  targeted_productivity       smv       wip  \
team                   1.000000               0.030274 -0.110011 -0.033474   
targeted_productivity  0.030274               1.000000 -0.069489  0.062054   
smv                   -0.110011              -0.069489  1.000000 -0.037837   
wip                   -0.033474               0.062054 -0.037837  1.000000   
over_time             -0.096737              -0.088557  0.674887  0.022302   
incentive             -0.007674               0.032768  0.032629  0.167210   
idle_time              0.003796              -0.056181  0.056863 -0.026299   
idle_men               0.026974              -0.053818  0.105901 -0.048718   
no_of_style_change    -0.011194              -0.209294  0.315388 -0.072357   
no_of_workers         -0.075113              -0.084288  0.912176  0.030383   
actual_productivity   -0.148753               0.421594 -0.122089  0.131147   

                       over_time  incentive  idle_time  idle_men  \
team                   -0.096737  -0.007674   0.003796  0.026974   
targeted_productivity  -0.088557   0.032768  -0.056181 -0.053818   
smv                     0.674887   0.032629   0.056863  0.105901   
wip                     0.022302   0.167210  -0.026299 -0.048718   
over_time               1.000000  -0.004793   0.031038 -0.017913   
incentive              -0.004793   1.000000  -0.012024 -0.021140   
idle_time               0.031038  -0.012024   1.000000  0.559146   
idle_men               -0.017913  -0.021140   0.559146  1.000000   
no_of_style_change      0.059790  -0.026607  -0.011598  0.133632   
no_of_workers           0.734164   0.049222   0.058049  0.106946   
actual_productivity    -0.054206   0.076538  -0.080851 -0.181734   

                       no_of_style_change  no_of_workers  actual_productivity  
team                            -0.011194      -0.075113            -0.148753  
targeted_productivity           -0.209294      -0.084288             0.421594  
smv                              0.315388       0.912176            -0.122089  
wip                             -0.072357       0.030383             0.131147  
over_time                        0.059790       0.734164            -0.054206  
incentive                       -0.026607       0.049222             0.076538  
idle_time                       -0.011598       0.058049            -0.080851  
idle_men                         0.133632       0.106946            -0.181734  
no_of_style_change               1.000000       0.327787            -0.207366  
no_of_workers                    0.327787       1.000000            -0.057991  
actual_productivity             -0.207366      -0.057991             1.000000

Targeted Productivity has a moderate positive correlation with actual productivity (0.42), meaning higher targets slightly boost actual performance.

SMV (task complexity) is strongly correlated with number of workers (0.91) and overtime (0.67), showing that complex tasks need more labor and time.

Idle time and idle workers negatively impact actual productivity (correlations of -0.08 and -0.18, respectively).
Style changes also reduce productivity (-0.21), as frequent changes cause disruptions.

Overtime has a weak negative effect on actual productivity (-0.05), indicating it may not improve outcomes efficiently.

In [13]:
# Check for duplicate rows
df.duplicated().sum()

0

df.duplicated().sum() counts the total number of duplicate rows in the DataFrame

In [14]:
# Summary of the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1197 entries, 0 to 1196
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   date                   1197 non-null   object 
 1   quarter                1197 non-null   object 
 2   department             1197 non-null   object 
 3   day                    1197 non-null   object 
 4   team                   1197 non-null   int64  
 5   targeted_productivity  1197 non-null   float64
 6   smv                    1197 non-null   float64
 7   wip                    691 non-null    float64
 8   over_time              1197 non-null   int64  
 9   incentive              1197 non-null   int64  
 10  idle_time              1197 non-null   float64
 11  idle_men               1197 non-null   int64  
 12  no_of_style_change     1197 non-null   int64  
 13  no_of_workers          1197 non-null   float64
 14  actual_productivity    1197 non-null   float64
dtypes: f

df.info() provides a concise summary of the DataFrame